art/defences/preprocessor/cutmix/cutmix_pytorch.py

```diff
@@ -0,0 +1,201 @@
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Optional, Tuple, TYPE_CHECKING
+
+import numpy as np
+from tqdm.auto import tqdm
+
+from art.defences.preprocessor.preprocessor import PreprocessorPyTorch
+
+if TYPE_CHECKING:
+    # pylint: disable=C0412
+    import torch
+
+logger = logging.getLogger(__name__)
+
+
+class CutMixPyTorch(PreprocessorPyTorch):
+    """
+    Implement the CutMix data augmentation defence approach in PyTorch.
+
+    | Paper link: https://arxiv.org/abs/1905.04899
+
+    | Please keep in mind the limitations of defences. For more information on the limitations of this defence,
+        see https://arxiv.org/abs/1803.09868 . For details on how to evaluate classifier security in general, see
+        https://arxiv.org/abs/1902.06705
+    """
+
+    params = ["num_classes", "alpha", "probability", "channels_first", "verbose"]
+
+    def __init__(
+        self,
+        num_classes: int,
+        alpha: float = 1.0,
+        probability: float = 0.5,
+        channels_first: bool = False,
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+        device_type: str = "gpu",
+        verbose: bool = False,
+    ) -> None:
+        """
+        Create an instance of a CutMix data augmentation object.
+
+        :param num_classes: The number of classes used for one-hot encoding.
+        :param alpha: The hyperparameter for sampling the combination ratio.
+        :param probability: The probability of applying CutMix per sample.
+        :param channels_first: Set channels first or last.
+        :param apply_fit: True if applied during fitting/training.
+        :param apply_predict: True if applied during predicting.
+        :param device_type: Type of device on which the classifier is run, either `gpu` or `cpu`.
+        :param verbose: Show progress bars.
+        """
+        super().__init__(
+            device_type=device_type,
+            is_fitted=True,
+            apply_fit=apply_fit,
+            apply_predict=apply_predict,
+        )
+        self.num_classes = num_classes
+        self.alpha = alpha
+        self.probability = probability
+        self.channels_first = channels_first
+        self.verbose = verbose
+        self._check_params()
+
+    def forward(
+        self, x: "torch.Tensor", y: Optional["torch.Tensor"] = None
+    ) -> Tuple["torch.Tensor", Optional["torch.Tensor"]]:
+        """
+        Apply CutMix data augmentation to sample `x`.
+
+        :param x: Sample to augment with shape of `NCHW`, `NHWC`, `NCFHW` or `NFHWC`.
+        :param y: Labels of `x` either one-hot or multi-hot encoded of shape `(nb_samples, nb_classes)`
+                  or class indices of shape `(nb_samples,)`.
+        :return: Data augmented sample. The returned labels will be probability vectors of shape
+                 `(nb_samples, nb_classes)`.
+        """
+        import torch  # lgtm [py/repeated-import]
```

Comment:
We have recently removed LGTM
```suggestion
        import torch
```